### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by radiansz**

### CODE REVIEW

File: client/components/ui/Input.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import React, { PropTypes } from 'react';
+
+
+const Input = ({ meta: { dirty, touched, error }, input: {onChange, ...restInput}, name, ...restProps }) => (
+  <div>
+    <input name={name} onChange={({target: {value}}) => { onChange(value) }} {...restInput} />
```
Comment (rosendi) [inline client/components/ui/Input.js]: Will it work? `<input {...input} />`

File: client/containers/ui/Form/Button.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+export class Button extends Component {
+  static contextTypes = {
+    formContext: PropTypes.object,
+  };
+
+  static propTypes = {
+    component: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.func,
```
Comment (rosendi) [inline client/containers/ui/Form/Button.js]: Why `PropTypes.func`? Could you please provide an example?

File: client/containers/ui/Form/Button.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import React, { Component, PropTypes } from 'react';
+
+export class Button extends Component {
+  static contextTypes = {
+    formContext: PropTypes.object,
```
Comment (rosendi) [inline client/containers/ui/Form/Button.js]: Please, use `shape`.

File: client/containers/ui/Form/Form.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
+import React, { Component, PropTypes } from 'react';
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: In general I think that `Form` should also export the necessary proptypes. See the comments below.

File: client/containers/ui/Form/Form.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+export default class Form extends Component {
+
+  static childContextTypes = {
+    formContext: PropTypes.object,
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Shape and export?

File: client/containers/ui/Form/Form.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  };
+
+  static propTypes = {
+    schema: PropTypes.object,
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Shape and export?

File: client/containers/ui/Form/Form.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  _onError(errors) {
+    const { onError } = this.props;
+
+    if (onError) {
+      onError(errors);
+    }
+  }
+
+  async _validate({ fields }) {
+    const fieldNames =  Object.keys(this.props.schema);
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]:
```
Object.keys().reduce((errors, fieldName, _, fieldNames) => {
  const await this._validateField({ name, value }) 
  return { ...errors, ...... }
```

Or something better :-)

File: client/containers/ui/Form/Form.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    schema: PropTypes.object,
+  };
+
+  state = {
+    fields: {},
+    errors: {},
+    submitting: false,
+  };
+
+  _onFieldChange(field, value) {
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Well, It's actually a public function. Maybe just `onChange`?

File: client/containers/ui/Form/Field.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        PropTypes.func,
+      ]).isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      value: props.value,
+      initialValue: props.value,
```
Comment (rosendi) [inline client/containers/ui/Form/Field.js]: Why is it in the state? It's just a prop, isn't it?

File: client/containers/ui/Form/Field.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      dirty: false,
+      touched: false,
```
Comment (rosendi) [inline client/containers/ui/Form/Field.js]:
```
submitting: false,
validating: false,
```

yep?

File: client/routes/signin/containers/SigninContainer.js [hunk 11]
```diff
@@ -1,6 +1,7 @@
 import { reduxForm } from 'redux-form';
 import SigninForm from '../components/SigninForm';
 import { signin } from '../modules/signin';
+import { connect } from 'react-redux';
 
 const formConfig = {
```
Comment (rosendi) [inline client/routes/signin/containers/SigninContainer.js]: Remove...

File: client/components/ui/Input.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import React, { PropTypes } from 'react';
+
+
+const Input = ({ meta: { dirty, touched, error }, input: {onChange, ...restInput}, name, ...restProps }) => (
```
Comment (rosendi) [inline client/components/ui/Input.js]: Base components: Input -> Text, TextArea, Checkbox, RadioGroup, CheckboxGroup. Thoughts?

File: client/containers/ui/Form/Button.js (hunk 2 above)
Comment (radiansz) [inline client/containers/ui/Form/Button.js]: Well, Input passed by SigninForm is been considered by React as a function, that's why i added it

File: client/containers/ui/Form/Form.js (hunk 6 above)
Comment (radiansz) [inline client/containers/ui/Form/Form.js]: It's kinda hard to describe that object via shape. Any kind of keys can be in this object, tho it's possible to specify custom validator with objectOf.

File: client/containers/ui/Form/Form.js (hunk 7 above)
Comment (radiansz) [inline client/containers/ui/Form/Form.js]: I dunno how to use await in such cases :cry:

File: client/containers/ui/Form/Field.js (hunk 9 above)
Comment (radiansz) [inline client/containers/ui/Form/Field.js]: fixed

File: client/components/ui/Input.js (hunk 1 above)
Comment (radiansz) [inline client/components/ui/Input.js]: We need to pass  value to onChange. In this case event would be passed

File: client/containers/ui/Form/Field.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+export default class Field extends Component {
+
+  static contextTypes = {
+    formContext: PropTypes.object.isRequired,
+  };
+
+  static propTypes = {
+    name: PropTypes.string.isRequired,
+    component: PropTypes.oneOfType([
```
Comment (rosendi) [inline client/containers/ui/Form/Field.js]: PropTypes.node.isRequired

File: client/containers/ui/Form/Field.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    return {
+      onChange: this.onChange,
+      onFocus: this.onFocus,
+      type,
+      value,
+    };
+  }
+
+  render() {
+    const { component: FieldComponent, name, ...restProps } = this.props;
```
Comment (rosendi) [inline client/containers/ui/Form/Field.js]: Order:  context, state, props

File: client/containers/ui/Form/Form.js [hunk 15]
```diff
@@ -0,0 +1,158 @@
+import React, { Component, PropTypes } from 'react';
+
+const propTypes = {
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Merge it with the class.

File: client/containers/ui/Form/Form.js [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+  schema: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.func)),
+  values: PropTypes.objectOf(PropTypes.any),
+  onError: PropTypes.func,
+  onSubmit: PropTypes.func,
+  children: PropTypes.node,
+};
+
+const childContextTypes = {
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Merge it with the props.

File: client/containers/ui/Form/Form.js [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  getChildContext() {
+    return {
+      formContext: {
+        fields: this.state.fields,
+        errors: this.state.errors,
+        values: this.props.values,
+        onFieldChange: this.onFieldChange,
+        submitting: this.state.submitting || false,
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Remove `|| false`

File: client/containers/ui/Form/Form.js [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+      ...this.state.fields,
+      [field]: value,
+    };
+
+    this._validate({ fields });
+
+    this.setState({ fields });
+  };
+
+  handleSubmit = (event) => {
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: `onSubmit`

File: client/containers/ui/Form/Form.js [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    this.submit();
+
+    event.preventDefault();
+  };
+
+  submit() {
+    const { onSubmit } = this.props;
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Order: state, props.

File: client/lib/validators/required.js [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  if (isValid) {
+    return null;
+  }
+
+  return message;
+};
+
+export default (...args) => new Promise(resolve => resolve(validator(...args)));
```
Comment (rosendi) [inline client/lib/validators/required.js]: `export default async (...args) => validator(...args));`

File: client/containers/ui/Form/Form.js [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+
+  _handleSubmitSuccess() {
+    this._endSubmitting();
+  }
+
+  _onError(errors) {
+    const { onError } = this.props;
+
+    if (onError) {
+      const submitErrors = onError(errors);
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Merge

File: client/containers/ui/Form/Form.js [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+  }),
+};
+
+export default class Form extends Component {
+
+  static childContextTypes = childContextTypes;
+
+  static propTypes = propTypes;
+
+  constructor(props) {
```
Comment (rosendi) [inline client/containers/ui/Form/Form.js]: Remove the constructor.

**Pull request merged**